### ISSUE TITLE
🚀 Feature Request: D1 Transaction support

### ISSUE THREAD

**Issue by codehz** [cloudflare/workers-sdk#2733]
### Describe the solution

Has the procedures feature mentioned in [this article](https://blog.cloudflare.com/whats-new-with-d1/) been shipped?
This feature does not seem to be available in the current version sdk, either locally or remotely. Is there another plan for this? I really want this feature.

One of my use cases is to cascade insert data into two (or more) tables, where the generated id obtained by the first insert statement is used as a parameter in the second insert statement. These two inserts need to be done in the same transaction, otherwise the inconsistent state may cause some problem. (In other database engines, such as postgresql, we can achieve this through subqueries or cte with returning clauses in one statement. Unfortunately, this method does not work in sqlite)

**Comment by rozenmd** [comment]
Hey @codehz - it hasn't been shipped yet, though @geelen will be working on it soon

**Comment by geelen** [comment]
Yeah afaik there's no way to do a subquery, but you can use `last_insert_rowid()` inside a batch call to get a bit of what you're after. I just confirmed this works in both native SQLite and D1:

```sql
PRAGMA foreign_keys=on;

drop table if exists b;
drop table if exists a;

create table a (
   id INTEGER PRIMARY KEY,
   value TEXT
);

create table b (
   id INTEGER PRIMARY KEY,
   a_id INTEGER,
   value TEXT,
   FOREIGN KEY (a_id) REFERENCES a(id)
);

INSERT INTO a (value) VALUES ('foo');
INSERT INTO b (a_id, value) VALUES (last_insert_rowid(), 'bar');

INSERT INTO a (value) VALUES ('baz');
INSERT INTO b (a_id, value) VALUES (last_insert_rowid(), 'boo');
```

<img width="801" alt="image" src="https://user-images.githubusercontent.com/23264/219142254-1594efc9-a5f4-466d-9a2c-b6670d5105dd.png">

Transactions _are_ coming but hopefully this can unblock you in the meantime?

**Comment by codehz** [comment]
It does work in this scenario, but in fact what I'm creating a generic framework that doesn't assume that users use rowid(or its alias) as a constraint for foreign key references.
Indeed, there is still another workaround to use the previous rowid in the subquery to query the primary key. This is exactly what I'm trying to do now.
it will looks like this:
```sql
create table A(id, value);
create table B(id, aid, value);
insert into A(id, value) values ('152fd082-842c-4f78-b9c1-21c02ab90d37', 42);
insert into B(id, aid, value) values ('92b723ac-4da1-465f-bafc-a56b89435afc', (select id from A where A.rowid = last_insert_rowid()), 8);
```
But this method is still not robust enough, because users may create [rowid-less tables](https://www.sqlite.org/withoutrowid.html) for some optimization purposes.

**Comment by geelen** [comment]
Yeah, it's definitely not a complete workaround, just a stopgap.

> I'm creating a [generic framework](https://github.com/codehz/qlite) that doesn't assume that users use rowid(or its alias) as a constraint for foreign key references.

I think you're going to have to disallow rowid-less tables for the moment, then. Then when true JS/SQL procedures land, you can make things more generic.

**Comment by codehz** [comment]
> I think you're going to have to disallow rowid-less tables for the moment, then. Then when true JS/SQL procedures land, you can make things more generic.

I'm going to disable the whole feature (Insert an object along with its related objects through relationships) right now, since it require more complex solution if I need insert more than one related objects (last_insert_rowid can only get last one. This means that it may be necessary to create a temporary table to store the rowids of these related objects. The benefits it brings are not worth compared to the cost of increased complexity, not to mention that it will soon be replaced by a new mechanism), and it's not a very useful feature either, just for completeness(since hasura do).

**Comment by geelen** [comment]
No worries. Hopefully this issue will help someone googling this problem in the short term, and I'll leave it open and update & close it once Transactions/Procedures lands.

**Comment by codehz** [comment]
BTW, in [this article](https://blog.cloudflare.com/whats-new-with-d1/), it is not clear which types we can pass to the Procedures 'worklet' (and return back). only primitive types? object which can be serialized to json (or [structuredClone](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone))? ArrayBuffer?

**Comment by geelen** [comment]
Can't be sure at this stage, but I'm aiming for full `structuredClone` support. Worst case JSON-serialisable only. 

If JSON is good enough though it might be easier to build so we might start there, ship that, then expand it later? Happy to hear suggestions on that front...

**Comment by codehz** [comment]
JSON's hidden problem is, it cannot contains cyclic reference(and this might bring some surprises for user since it is transparent), other than that, the json looks good enough

**Comment by LuckyWraptor** [comment]
Any ETA on this feature?

**Comment by AmaraFray** [comment]
Any Updates on this feature?

**Comment by FlareLine** [comment]
Hey @geelen - just checking in here - do we see transaction support being ready within the next couple of releases? 😁

**Comment by takakikasuga** [comment]
`+1`👀

**Comment by duducpp** [comment]
# 👀🆘

**Comment by richard-egeli** [comment]
👀

**Comment by mohamnag** [comment]
Feb 2023 to Dec 2024 and still nothing. its really a shame for D1 missing such an important feature!

**Comment by Ehbraheem** [comment]
This is a serious concern for us. Without transaction support, D1 is practically unsuitable for production OLTP usage

**Comment by rxliuli** [comment]
Are there any updates? Or can we only use batch submission?

**Comment by Nazvix** [comment]
We need this feature. Cant keep using Try and Catch exception, unreliable and blobby.

**Comment by ZhangPeng4242** [comment]
any updates?

**Comment by FlareLine** [comment]
@geelen @rozenmd @penalosa I'm keen to know what's happening here as someone who contributed to the D1 implementation in `workers-rs` - is this just a case of capacity? Do you need someone to work on this? Is there anything the open-source community can do to get this over the line?

From what I can understand, there's not really any way someone external to Cloudflare can pick this up outside of an NDA given it relates to intricate systems / workings of D1, but is there room for anyone to speed implementation of this feature up?

There are a non-insignificant number of people now relying on this project in production systems, including several downstream systems such as prism and drizzle, and provided the currently-avaialble workarounds aren't suitable in most instances it seems a bit strange that this has stagnated for as long as it has.

**Comment by vy-ton** [comment]
D1 Product Manager here. There are no current or future plans to build interactive Javascript transaction support for D1, which is a significant effort. The [past blog](https://blog.cloudflare.com/whats-new-with-d1/) described a proposal when D1 had a different underlying architecture, which was replaced for performance and reliability issues.


I would recommend folks check out [SQLite in Durable Objects](https://developers.cloudflare.com/durable-objects/changelog/#2024-09-26) if interactive Javascript transaction support is a requirement for any workload. By design, Durable Objects provide the intended code running alongside embedded database storage abstraction.

**Comment by inhwa-son** [comment]
+1

**Comment by Crsk** [comment]
@connorads 

Startbase and Outerbase are interfaces that connects to the db, not a db engine itself.

And since D1 transactions are not possible mainly due to a SQLite limitation allowing only a single transaction at once, means a workaround is doing transactions through Durable Objects, because each of them has their own SQLite instance.

In other words i don't think they'd support D1 transactions outside of Durable Objects.

**Comment by infodusha** [comment]
> due to a SQLite limitation allowing only a single transaction at once
There is for example libSQL that removes such limitation, so it is not about technical side of the issue...

**Comment by vinodhj** [comment]
👀

**Comment by samuelyung** [comment]
Can `batch()` statement be used as a workaround?

**Comment by Crsk** [comment]
> Can `batch()` statement be used as a workaround?

Of course, if you don't mind not locking DB which allows safe read-then-write operations.

SQLite supports only one open transaction at a time, then it makes sense why it's not supported, imagine each user able to lock the entire database...

Durable Objects creates many SQLite instances though, so no global lock, meaning closer workaround.

But yeah batch is ok if you don't mind leaving the door open for possible data integrity issues.

**Comment by verespej** [comment]
I recently found batch useful for some simple cases. So, I wanted to share in case this can help anyone else solve similar cases.

From the current [batch documentation](https://developers.cloudflare.com/d1/worker-api/d1-database/#batch):

> Our implementation guarantees that each statement in the list will execute and commit, sequentially, non-concurrently.
> 
> Batched statements are [SQL transactions ↗](https://www.sqlite.org/lang_transaction.html). If a statement in the sequence fails, then an error is returned for that specific statement, and it aborts or rolls back the entire sequence.


So, D1's batch should allow us to run a sequence of operations atomically. (Important to note that not all drivers behave this way. So, be careful not to assume it'll work for drivers other than D1.)

In my case, I use random IDs for entities (versus monotonically increasing IDs). So, I can use a manually-generated ID across dependent, sequential operations. Here's a contrived example:
```
const userId = crypto.randomUUID();
const userEmail = "test@example.com";

const taskId = crypto.randomUUID();
const taskTitle = "Do something";

const insertUser = env.DB.prepare(`
  INSERT INTO Users (id, email)
  VALUES (?, ?)
`).bind(userId, userEmail);

const insertTask = env.DB.prepare(`
  INSERT INTO Tasks (id, user_id, title)
  VALUES (?, ?, ?)
`).bind(taskId, userId, taskTitle);

const batchResult = await env.DB.batch([
  insertUser,
  insertTask,
]);
```

In this case, D1 should:
1. Create the User with our manually-generated ID
2. Create the Task with the newly-created user's ID
    1. If creation of the Task fails, roll back the creation of the User and return an error

This definitely doesn't solve all needs, but I've found it useful for some common, simple cases.

**Comment by hnordt** [comment]
@crsk

> But yeah batch is ok if you don't mind leaving the door open for possible data integrity issues.

Could you explain why using D1 batch might still cause data integrity issues? The docs say that batches are transactions, so if one operation fails, the others should be rolled back. Am I misunderstanding something?

**Comment by Crsk** [comment]
> [@Crsk](https://github.com/Crsk)
> 
> > But yeah batch is ok if you don't mind leaving the door open for possible data integrity issues.
> 
> Could you explain why using D1 batch might still cause data integrity issues? The docs say that batches are transactions, so if one operation fails, the others should be rolled back. Am I misunderstanding something?

Sure, so traditional transactions lock DB to ensure safe writes in a known consistent state, batch while atomic (all-or-nothing), doesn't lock DB.

If you have $100 and three people try to withdraw at the same time and one of the batch operations fails, the third rolls back to $0 but the first two still get $200 out of $100.

**Comment by hnordt** [comment]
@Crsk am I safe as long as I don’t need a read-then-write? I mean, `batch()` only guarantees that all statements in the batch either all succeed or all fail, nothing more. Right?

**Comment by Crsk** [comment]
> [@Crsk](https://github.com/Crsk) am I safe as long as I don’t need a read-then-write? I mean, `batch()` only guarantees that all statements in the batch either all succeed or all fail, nothing more. Right?

yep, that's right

**Comment by a-eid** [comment]
wondering if there is any update about this feature request... it's been quit a while

**Comment by NoirHare** [comment]
Is there any progress on this feature?

**Comment by cyraid** [comment]
@Crsk I'm not sure I understand. With this batch (all or nothing), if all 3 people try to take out from the $100, how much are they taking out? By you saying $200, I'm guessing they're trying to take $100 out. But how are 2 end up with $100 each? If it's all or nothing, wouldn't that transaction have failed? Can you show some test code of what you mean? And can I reproduce this?

**Comment by Crsk** [comment]
> [@Crsk](https://github.com/Crsk) I'm not sure I understand. With this batch (all or nothing), if all 3 people try to take out from the $100, how much are they taking out? By you saying $200, I'm guessing they're trying to take $100 out. But how are 2 end up with $100 each? If it's all or nothing, wouldn't that transaction have failed? Can you show some test code of what you mean? And can I reproduce this?

Two users can see the same $100 and update it, unaware of each other. A real transaction locks DB as soon as other touches it, so the second has to wait until the first finishes.

```SQL
-- Batch
UPDATE account SET balance = balance - 100 WHERE id = 1;
UPDATE account SET balance = balance - 100 WHERE id = 1;
-- Both run at the same time, final balance = -100

-- Transaction
BEGIN;
UPDATE account SET balance = balance - 100 WHERE id = 1;
COMMIT;

-- User B must wait until User A COMMITs
BEGIN;
UPDATE account SET balance = balance - 100 WHERE id = 1;
COMMIT;
-- Runs sequentially, no double spend
```

Of course, you can always add a constraint like:
```SQL
ALTER TABLE account ADD CONSTRAINT balance_nonnegative CHECK (balance >= 0);
```
but that only stops negative values. It doesn’t solve the batch issue.

Atomicity (all or nothing) doesn't mean isolation (safe read-then-write). Batch doesn't lock.

**Comment by cyraid** [comment]
@Crsk But the article states:
> .batch() (which does in fact use database transactions under the hood)

The problem (as far as I understand/read), is that when opening up a Write Lock, the entire Database (not the nearest replica close to the user which are for reads) would lock the DB, and if a Worker were to crash without a rollback for example, would block the whole system.

It sounds like the author of the article makes it seem like locking the whole DB would be a bad thing. While I somewhat agree, I do think there should at least be a method to do so (even if there are timeouts that could be in place incase something drastic were to happen). Perhaps even a lock table.

**Comment by a-eid** [comment]
is this featured planned at all ?

**Comment by penalosa** [comment]
Based on [@vy-ton's comment](https://github.com/cloudflare/workers-sdk/issues/2733#issuecomment-2712365336), I'm going to close this
